_posts/2017-07-12-TypeScript-for-JS-devs.markdown

```diff
@@ -0,0 +1,485 @@
+  role: "Engineer",
+  joined: "2 years ago",
+  floor: "25"
+}
+
+```
+
+You might wonder how this fits with ES6 classes? You can use classes as at Type. TypeScript classes are not much different then the first class language construct in Traditional JS. In TypeScript you can use classes to define a specific type.
+
+```ts
+interface Point {
+  x: number 
+  y: number
+}
+
+class MyPoint implements Point {
+   x: number; y: number; 
+}
+
+let point: MyPoint = new MyPoint()
+point.x = 3;
+point.y = 8;
+
+point.y = "10"; // compile error
+```
+
+# Generics
+
+
+
+TypeScript (especially code we write with React) heavily uses generics to handle related classes. 
+
+In this case, we’re saying we want an Array with an associated type of string. This allows TypeScript to know what type of object would be returned from a get function.
+
+```ts
+const names: Array<string> = [
+  "christina", 
+  "orta",
+  "danger"
+]
+```
+
+Without generics, you’d have to create a version of the array that only returns a specific type. Instead you would say “this is an array - and everything inside it is going to be a string” [...]
+
+```ts
+const names: ArrayString = [
+  "christina", 
+  "orta",
+  "danger"
+]
+```
+
+There is shorthand for this in the form of `array[]`.
+
+## Generic Promises
+
+Let’s look at another example - Promises
+
+In this case we are stating that the promise will return a number. This associates the returned value inside the promise with the API. It means we would get compiler errors if that is treated as a string.
+
+```ts
+const getAge:Promise<number> = new Promise(res => 5)
+
+const age = await getAge()
+if (age === "18") {
+  ...
+}
+```
+
+## Generic Functions
+
+These are useful for when you have a function which allows you to work with many types, and keep them consistent. 
+
+For example this reverse array function will take a type called T, which comes from the param and it will return an array of the same type
+
+This is generally only useful for utilities funds, classes or libraries
+
+```ts
+function reverseArray<T>(arr:T[]): T[] {
+  return arr.reverse()
+}
+
+const nums = [1, 2, 3]
+console.log(reverseArray(nums)) // [3, 2, 1]
+
+const strs = ["1", "2", "3"]
+console.log(reverseArray(strs)) // ['3', '2', '1']
+```
+
+## Generics in Production
+
+Speaking of which, this is the definition of a React component. Let’s clean this up.
+
+```js
+// Base component for plain JS classes
+class Component<P, S> implements ComponentLifecycle<P, S> {
+    constructor(props?: P, context?: any);
+    setState<K extends keyof S>(f: (prevState: S, props: P) => Pick<S, K>, callback?: () => any): void;
+    setState<K extends keyof S>(state: Pick<S, K>, callback?: () => any): void;
+    forceUpdate(callBack?: () => any): void;
+    render(): JSX.Element | null;
+
+    props: Readonly<{ children?: ReactNode }> & Readonly<P>;
+    state: Readonly<S>;
+    context: any;
+    refs: {
+        [key: string]: ReactInstance
+    };
+}
+```
```

Comment:
From a general reader perspective, is this section necessary? Unless the paragraph below intends to step through in detail this `Generics in Production` section seems to just overwhelm the overall post.